src/main/scala/org/ergoplatform/nodeView/wallet/TrackedBoxSerializer.scala

```diff
@@ -0,0 +1,214 @@
+
+import scala.util.{Failure, Try}
+
+class TrackedBoxSerializer(txLookup: TransactionLookup) extends TypedBoxSerializer[TrackedBox] {
+
+  def write(trackedBox: TrackedBox, w: ByteWriter): Unit = {
+    trackedBox match {
+      case b: UnspentOffchainBox => new UnspentOffchainBoxSerializer(txLookup).write(b, w)
+      case b: UnspentOnchainBox => new UnspentOnchainBoxSerializer(txLookup).write(b, w)
+      case b: SpentOffchainBox => new SpentOffchainBoxSerializer(txLookup).write(b, w)
+      case b: SpentOnchainBox => new SpentOnchainBoxSerializer(txLookup).write(b, w)
+    }
+  }
+
+  override def read(r: ByteReader): Try[TrackedBox] = {
+    val pos = r.position
+    val (_, spendingStatus, onchainStatus) = readHeaderBits(r)
+    r.position = pos
+
+    (spendingStatus, onchainStatus) match {
+      case (Unspent, Offchain) => new UnspentOffchainBoxSerializer(txLookup).read(r)
+      case (Unspent, Onchain) => new UnspentOnchainBoxSerializer(txLookup).read(r)
+      case (Spent, Offchain) => new SpentOffchainBoxSerializer(txLookup).read(r)
+      case (Spent, Onchain) => new SpentOnchainBoxSerializer(txLookup).read(r)
+    }
+  }
+
+}
+
+object TrackedBoxSerializer {
+  type TransactionLookup = ModifierId => Option[ErgoTransaction]
+}
+
+class UnspentOffchainBoxSerializer(txLookup: TransactionLookup) extends TypedBoxSerializer[UnspentOffchainBox] {
+
+  def write(unspentOffchainBox: UnspentOffchainBox, w: ByteWriter): Unit = {
+    import unspentOffchainBox._
+    w.putBits(headerBits(unspentOffchainBox))
+      .putBytes(idToBytes(creationTx.id))
+      .putShort(creationOutIndex)
+    ErgoBoxSerializer.write(box, w)
+  }
+
+  def read(r: ByteReader): Try[UnspentOffchainBox] = {
+    readHeader(r, Unspent, Offchain, "UnspentOffchainBox"){ certainty =>
+      readTx(r, txLookup) { tx =>
+        val outIndex = r.getShort()
+        readErgoBox(r) { box =>
+          UnspentOffchainBox(tx, outIndex, box, certainty)
+        }
+      }
+    }
+  }
+}
+
+class UnspentOnchainBoxSerializer(txLookup: TransactionLookup) extends TypedBoxSerializer[UnspentOnchainBox] {
+
+  def write(unspentOnchainBox: UnspentOnchainBox, w: ByteWriter): Unit = {
+    import unspentOnchainBox._
+    w.putBits(headerBits(unspentOnchainBox))
+      .putBytes(idToBytes(creationTx.id))
+      .putShort(creationOutIndex)
+      .putInt(creationHeight)
+    ErgoBoxSerializer.write(box, w)
+  }
+
+  def read(r: ByteReader): Try[UnspentOnchainBox] = {
+    readHeader(r, Unspent, Onchain, "UnspentOnchainBox"){ certainty =>
+      readTx(r, txLookup) { tx =>
+        val outIndex = r.getShort()
+        val height = r.getInt()
+        readErgoBox(r) { box =>
+          UnspentOnchainBox(tx, outIndex, height, box, certainty)
+        }
+      }
+    }
+  }
+
+}
+
+class SpentOffchainBoxSerializer(txLookup: TransactionLookup) extends TypedBoxSerializer[SpentOffchainBox] {
+
+  def write(spentOffchainBox: SpentOffchainBox, w: ByteWriter): Unit = {
+    import spentOffchainBox._
```

Comment:
this import looks quite strange. I think it would be much cleaner to write `spentOffchainBox.creationTx` and so on for class fields

Reply:
This was to make code looking native to Box and to avoid "train wreck" call chaining like `spentOffchainBox.creationTx.id`. 

But ok, I'll to remove this